src/System.Runtime.CompilerServices.Unsafe/src/System.Runtime.CompilerServices.Unsafe.il

```diff
@@ -281,6 +281,20 @@
 #endif
   } // end of method Unsafe::AsRef
 
+  .method public hidebysig static !!T& AsRef<T>(!!T& source) cil managed aggressiveinlining
```

Comment:
Isn't this supposed to also have a `modreq`: https://github.com/dotnet/csharplang/blob/ac995bd794be5a17acb52c30dfa4d61a1bfd15c9/proposals/csharp-7.2/readonly-ref.md#metadata-representation?

Reply:
modregs are used on returned `ref readonly` only

Reply:
Doesn't that make it impossible to have both ref and ref read-only overloads, since the would compile down to the same il signature? As far as I know, the latest decision was to allow both ovetloads and only  require the `in` keyword if the decision was ambiguous.

Reply:
In order to make this change work it has to match what the current compiler does. If there are plans to keep tweaking the implementation, this will need to be updated to match. It is common for  compiler upgrade in corefx to require accompanying source changes.

Reply:
Looking at the current bits/tests in master, a modreq is required: https://github.com/dotnet/roslyn/blob/1fd9a04c8c0d13afdd7af546fa7dce83549d6aba/src/Compilers/CSharp/Test/Emit/Emit/InAttributeModifierTests.cs#L205 (others as well, if you just search for modreq).

@vsadov or @jcouv could probably comment more.

Reply:
That's fine - these changes are not in the compiler that is used to compile corefx. This just means that this will need to be updated to match on next compiler update in corefx.

Reply:
It would be useful if (before merging) we had a bug tracking the appropriate fix when next the compiler is upgraded.

Reply:
I do not think modreq is required in this case. It is not a return or a parameter of something overridable.
Try compiling similar signature in c# just in case.

Reply:
Compiling the signature in C# doesn't seem to generate a modreq; see https://github.com/dotnet/corefx/pull/24479#issuecomment-335008568.

Reply:
@VSadov, I would fully expect the modreq to be required, otherwise you cannot disambiguate (from the IL method signatures alone) whether it is `ref` or `ref readonly` (and therefore cannot overload between the two).

The C# signature, as it is currently defined, would be: `public static ref T AsRef<T>([IsReadOnly] ref T source)`. Given that the attribute is not part of the IL signature, it is not-distinguishable from `public static ref T AsRef<T>(ref T source)` (both compile to an IL signature of `.method public hidebysig static !!T& AsRef<T>(!!T& source) cil managed aggressiveinlining`.

Reply:
@VSadov, I checked the C# behavior in current master and decided to send a mail.

I think there are some clear issues with what we are currently doing and some not so clear (or just plain missing) documents on what the final expected behavior is.

From a basic standpoint:
* Converting an existing method from `ref` to `in` should not be breaking (it currently is)
  * This only applies if you can't have both `ref` and `in` overloads (as it is currently)
* The marshaller should not need to be updated in order to work with `in` parameters
  * The compiler should be emitting the `[In]` attribute, as it does the `[Out]` attribute for `out` parameters

Reply:
@tannergooding I believe there is some confusion here. The modreq is not used to distinguish between `in` parameters and other ref kinds. The attribute `IsReadOnly` is used for that purpose (and emitted in all cases. The modreq is only used in cases where we want to prevent older compilers/other languages to compile implementations/overrides without understanding the semantics. Am I missing something?